WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs

```diff
@@ -72,6 +90,53 @@ public Pocket[] AutoSelectPockets()
 		return _allPockets.ToArray();
 	}
 
+	private Pocket[]? GetBestKnownByRecipientPockets(Pocket[] knownByRecipientPockets, Money targetAmount, SmartLabel recipient)
+	{
+		var privacyRankedPockets =
+			knownByRecipientPockets
+				.Select(pocket =>
+				{
+					var containedRecipientLabelsCount = pocket.Labels.Count(label => recipient.Contains(label, StringComparer.OrdinalIgnoreCase));
+					var totalPocketLabelsCount = pocket.Labels.Count();
+					var totalRecipientLabelsCount = recipient.Count();
+					var index = ((double)containedRecipientLabelsCount / totalPocketLabelsCount) + ((double)containedRecipientLabelsCount / totalRecipientLabelsCount);
+
+					return (acceptabilityIndex: index, pocket);
+				})
+				.OrderByDescending(tup => tup.acceptabilityIndex)
+				.ThenBy(tup => tup.pocket.Labels.Count())
+				.ThenByDescending(tup => tup.pocket.Amount)
+				.Select(tup => tup.pocket)
+				.ToArray();
+
+		var pockets = new List<Pocket>();
+		foreach (var p in privacyRankedPockets)
+		{
+			if (pockets.Sum(x => x.Amount) < targetAmount)
+			{
+				pockets.Add(p);
+			}
+			else
+			{
+				break;
+			}
+		}
+
+		foreach (var p in pockets.OrderBy(x => x.Amount).ToImmutableArray())
+		{
+			if (pockets.Sum(x => x.Amount) - p.Amount >= targetAmount)
+			{
+				pockets.Remove(p);
+			}
+			else
+			{
+				break;
+			}
+		}
```

Comment:
In the `privacyRankedPockets` you have pockets [A,B,C,D,E].
You start to select pockets from the first to the last from that list until they cover the target amount.
Let's say A, B, and C cover it. But it can happen that B and C are enough and A is unnecessary. So here I order the pockets by amounts `pockets.OrderBy(x => x.Amount).ToImmutableArray()` and check if the remaining `pockets` without `p` does still cover the target amount. If so then `p` is not necessary.
So continue the example... pocket A is the best from the privacy perspective but it is not necessary, so here it will be removed. 

And the result is B and C. And we also know that D and E are worse from the privacy perspective so we don't have to deal with them. 

I will add test to this case.

Reply:
> So here I order the pockets by amounts `pockets.OrderBy(x => x.Amount).ToImmutableArray()` 

I think we should order by amounts AND also by number of labels, so if the amounts of A and B are equal then we should take A and remove B.

`foreach (var p in bestPockets.OrderBy(x => x.Amount).ThenByDescending(x => x.Labels.Count()).ToImmutableArray())`

If you agree with that then we can edit `AutoSelectRequiredKnownByRecipientPockets` as follow:

```
public void AutoSelectRequiredKnownByRecipientPockets()
{
	var selection = new LabelSelectionViewModel(Money.Parse("1.0"));

	var pockets = new List<Pocket>();
	pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
	pockets.AddPocket(1.1M, out var pocket2, "Dan");
	pockets.AddPocket(1.5M, out var pocket3, CoinPocketHelper.UnlabelledFundsText);
	pockets.AddPocket(0.4M, out var pocket4, "David");
	pockets.AddPocket(0.4M, out var pocket5, "David", "Max");
	pockets.AddPocket(0.6M, out var pocket6, "David", "Lucas", "Dan");

	selection.Reset(pockets.ToArray());

	var output = selection.AutoSelectPockets("David");
	Assert.DoesNotContain(pocket1, output);
	Assert.DoesNotContain(pocket2, output);
	Assert.DoesNotContain(pocket3, output);
	Assert.Contains(pocket4, output);
	Assert.DoesNotContain(pocket5, output);
	Assert.Contains(pocket6, output);
}
```

Reply:
Done